README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# mapiso <img src="man/figures/logo.png" align="right" width="140"/>

<!-- badges: start -->

[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![CRAN](https://www.r-pkg.org/badges/version/mapiso)](https://cran.r-project.org/package=mapiso)
[![R-CMD-check](https://github.com/riatelab/mapiso/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/riatelab/mapiso/actions/workflows/R-CMD-check.yaml)
[![Codecov test
coverage](https://codecov.io/gh/riatelab/mapiso/branch/main/graph/badge.svg)](https://app.codecov.io/gh/riatelab/mapiso?branch=main)
<!-- badges: end -->

The goal of `mapiso` is to ease the transformation of regularly spaced
grids containing continuous data into contour polygons. These grids can
be defined by data.frames (x, y, value), `sf` objects or SpatRasters
from `terra`.  
`mapsio` is a wrapper around [`isoband`](https://isoband.r-lib.org).

## Installation

You can install the released version of `mapiso` from
[CRAN](https://cran.r-project.org/package=mapiso) with:

``` r
install.packages("mapiso")
```

Alternatively, you can install the development version of `mapiso` from
GitHub with:

``` r
remotes::install_github("riatelab/mapiso")
```

## Usage

### Raster

``` r
library(mapiso)
library(terra)
#> terra 1.7.29
library(mapsf)
r <- rast(system.file("tif/elevation.tif", package = "mapiso"))
isor <- mapiso(x = r)
mf_theme(mar = c(0, 0, 0, 0))
mf_raster(r)
mf_map(isor, col = NA, add = TRUE)
```

<img src="man/figures/README-raster-1.png" width="80%" />

### sf regular grid

``` r
library(mapiso)
library(sf)
#> Linking to GEOS 3.9.0, GDAL 3.2.2, PROJ 7.2.1; sf_use_s2() is TRUE
library(mapsf)
# gridded data
s <- st_read(system.file("gpkg/elevation.gpkg", package = "mapiso"), 
             layer = "elevation", quiet = TRUE)
# mask
m <- st_read(system.file("gpkg/elevation.gpkg", package = "mapiso"),
             layer = "com", quiet = TRUE)
# custom breaks
bks <-c(98,100, 150, 200, 250, 300, 350, 400, 412.6) 
isos <- mapiso(x = s, var = "elevation", breaks = bks, mask = m)
mf_map(isos, "isomin", "choro", 
       breaks = bks, border = NA, 
       leg_title = "elevation")
mf_map(m, col = NA, add = TRUE)
mf_map(s, cex = 1, pch = ".", col = "grey20", add = TRUE)
```

<img src="man/figures/README-sf-1.png" width="80%" />

### data.frame

``` r
library(mapiso)
library(mapsf)
d <- read.csv(system.file("csv/elevation.csv", package = "mapiso"))
head(d)
#>          x       y elevation
#> 1 586231.3 6431677  282.0299
#> 2 586481.3 6431677  263.7194
#> 3 586731.3 6431677  244.9597
#> 4 586981.3 6431677  222.3232
#> 5 587231.3 6431677  211.1522
#> 6 587481.3 6431677  232.1604
isod <- mapiso(x = d, var = 'elevation', coords = c('x', 'y'), crs = 'epsg:2154')
bks <- unique(c(isod$isomin, isod$isomax))
mf_map(isod, "isomin", "choro", breaks = bks, leg_title = "elevation")
```

<img src="man/figures/README-data.frame-1.png" width="80%" />

## Community Guidelines

One can contribute to the package through [pull
requests](https://github.com/riatelab/mapiso) and report issues or ask
questions [here](https://github.com/riatelab/mapiso/issues).  
This project uses [conventional
commits](https://www.conventionalcommits.org/en/v1.0.0-beta.3/) and
[semantic versioning](https://semver.org/).


NEWS.md
# mapiso 0.3.0

## Fix
* better min and max values, avoid rounding problems for upper values


# mapiso 0.2.0

## Fix
* reorder data.frame and sf grids along x and y coordinates

## Feat
* add support for SpatVector (input/output)



# mapiso 0.1.2 

## Fix
- avoid invalid geometries by shifting values 
- remove error when CRS do not match between x and mask


tests/tinytest.R
if (requireNamespace("tinytest", quietly = TRUE)) {
  library(tinytest)
  test_package("mapiso")
}


inst/tinytest/test_mapiso.R
library(tinytest)
suppressPackageStartupMessages(library(sf))
s <- st_read(system.file("gpkg/elevation.gpkg", package = "mapiso"),
             layer = "elevation", quiet = TRUE)
m <- st_read(system.file("gpkg/elevation.gpkg", package = "mapiso"),
             layer = "com", quiet = TRUE)
d <- read.csv(system.file("csv/elevation.csv", package = "mapiso"))
bks <-c(98,100, 150, 200, 250, 300, 350, 400, 412.6)



if(suppressPackageStartupMessages(require(terra, quietly = TRUE))){
  r <- rast(system.file("tif/elevation.tif", package = "mapiso"))
  expect_silent(mapiso(x = r))
  r$coco <- r$elevation +1
  expect_error(mapiso(r))
  expect_silent(mapiso(x = vect(s), var = "elevation",
                       breaks = bks, mask = vect(m)))

}

expect_silent(mapiso(x = s, var = "elevation", breaks = bks, mask = m))
expect_silent(mapiso(x = s, var = "elevation", breaks = bks))
expect_silent(mapiso(x = d, var = 'elevation', coords = c('x', 'y'),
                     crs = 'epsg:2154'))



# test errors
expect_error(mapiso("textx"))
expect_silent(mapiso(x = s, var = "elevation", breaks = bks,
                     mask = st_transform(m, 4326)))
expect_error(mapiso(s, var = "nope"))
expect_error(mapiso(s))
expect_error(mapiso(s[-1, ], var = "elevation"))
expect_error(mapiso(d))
expect_error(mapiso(d, var = "elevation"))
expect_error(mapiso(d, var = "nope", coords = c("x", "y")))


expect_equal(mapiso(x = s, var = "elevation", breaks = bks),
             mapiso(x = s[order(s$elevation), ],
                    var = "elevation", breaks = bks))

expect_equal(mapiso(x = d, var = 'elevation', coords = c('x', 'y'),
                    crs = 'epsg:2154'),
             mapiso(x = d[order(d$elevation), ], var = 'elevation',
                    coords = c('x', 'y'),
                    crs = 'epsg:2154'))


data-raw/build_dataset.R
library(terra)
ra <- terra::rast(system.file("tif/elev.tif", package = "tanaka"))
names(ra) <- "elevation"
ra2 <- ra
res(ra2) <- 250
a <- resample(ra, ra2)
terra::writeRaster(a, "inst/tif/elevation.tif", overwrite = T)
ra <- terra::rast("inst/tif/elevation.tif")
df <- as.data.frame(ra, xy = T)
write.csv(df, "inst/csv/elevation.csv", row.names = F)
df <- read.csv("inst/csv/elevation.csv")
sdf <- sf::st_as_sf(df, coords = c("x", "y"), crs = 2154)
sf::st_write(sdf, "inst/gpkg/elevation.gpkg", append = FALSE)
com <- sf::st_read(system.file("gpkg/com.gpkg", package = "tanaka"),
  quiet = TRUE
)
sf::st_write(com, "inst/gpkg/elevation.gpkg", layer = "com", append = FALSE)



## Build logo

library(mapiso)
library(sf)
library(mapsf)
# gridded data
s <- st_read(system.file("gpkg/elevation.gpkg", package = "mapiso"),
  layer = "elevation", quiet = TRUE
)
# mask
m <- st_read(system.file("gpkg/elevation.gpkg", package = "mapiso"),
  layer = "com", quiet = TRUE
)
# custom breaks
bks <- c(98, 100, 150, 200, 250, 300, 350, 400, 412.6)
isos <- mapiso(x = s, var = "elevation", breaks = bks)
mf_export(m, "toto.svg")
mf_map(isos, "isomin", "choro",
  add = T,
  breaks = bks, border = NA, leg_pos = NA,
  leg_title = "elevation", pal = "Sunset"
)
mf_map(s, "elevation", "prop",
  leg_pos = NA,
  col = "#ffffff80", inches = .05, border = "#00000050"
)
mf_map(isos, add = T, border = "grey30", col = NA, lwd = .5)

dev.off()


R/mapiso.R
#' @title Create Contour Polygons from Regular Grids
#' @description Regularly spaced grids containing continuous data are
#' transformed into contour polygons. A grid can be defined by a
#' data.frame (x, y, value), an \code{sf} object, a \code{terra} SpatRaster or
#' SpatVector.
#'
#' @param x a data.frame, an sf object or a SpatRaster
#' @param var name of the variable, for data.frames and sf objects only
#' @param breaks list of break values (default to equal interval)
#' @param nbreaks number of classes
#' @param mask an sf object or SpatVector of polygons or multipolygons.
#' \code{mask} is  used to clip contour polygons
#' @param coords names of the coordinates variables
#' (e.g. \code{c("lon", "lat")}), for data.frames only
#' @param crs CRS code (e.g. "epsg:2154"), for data.frames only.
#' @importFrom sf st_union st_intersection st_cast st_agr<- st_coordinates
#' st_crs st_geometry st_sf st_sfc st_as_sf
#' @importFrom isoband isobands iso_to_sfg
#' @return The output is an sf object of polygons (or a SpatVector if \code{x}
#' is a SpatVector). The data.frame contains three fields:
#' id (id of each polygon), isomin and isomax (minimum and maximum
#' breaks of the polygon).
#' @export
#' @examples
#' # sf, using a mask
#' library(sf)
#' s <- st_read(system.file("gpkg/elevation.gpkg", package = "mapiso"),
#'   layer = "elevation", quiet = TRUE
#' )
#' m <- st_read(system.file("gpkg/elevation.gpkg", package = "mapiso"),
#'   layer = "com", quiet = TRUE
#' )
#' isos <- mapiso(
#'   x = s, var = "elevation",
#'   mask = m
#' )
#' plot(isos)
#'
#' # data.frame, using user breaks values
#' d <- read.csv(system.file("csv/elevation.csv", package = "mapiso"))
#' bks <- c(98, 100, 150, 200, 250, 300, 350, 400, 412.6)
#' isod <- mapiso(
#'   x = d, var = "elevation",
#'   breaks = bks, coords = c("x", "y"), crs = "epsg:2154"
#' )
#' plot(isod)
#' if (require(mapsf, quietly = TRUE)) {
#'   mf_map(isod, "isomin", "choro", breaks = bks, leg_title = "Elevation")
#' }
#' \dontrun{
#' if (require(terra, quietly = TRUE)) {
#'   # terra SpatRaster
#'   r <- rast(system.file("tif/elevation.tif", package = "mapiso"))
#'   isor <- mapiso(x = r)
#'   plot(r)
#'   plot(st_geometry(isor), add = TRUE, col = NA)
#'   # terra SpatVector
#'   s_terra <- vect(s)
#'   m_terra <- vect(m)
#'   isost <- mapiso(
#'     x = s_terra, var = "elevation", mask = m_terra
#'   )
#'   plot(isost)
#' }
#' }
#'
mapiso <- function(x, var, breaks, nbreaks = 8, mask, coords, crs) {
  sptvct <- FALSE
  # test inputs
  if (!inherits(x = x, what = c(
    "SpatRaster", "SpatVector",
    "sf", "data.frame"
  ))) {
    stop(
      paste0(
        "'x' should be a data.frame, an sf data.frame a SpatVector ",
        "or a SpatRaster."
      ),
      call. = FALSE
    )
  }

  if (inherits(x = x, what = c("SpatRaster", "SpatVector"))) {
    if (!requireNamespace("terra", quietly = TRUE)) {
      stop(
        paste0(
          "This function needs the 'terra' package to work with ",
          "'terra' objects. Please install it."
        ),
        call. = FALSE
      )
    }
    if (inherits(x = x, "SpatVector")) {
      sptvct <- TRUE
      x <- st_as_sf(x)
    }
    if (inherits(x = x, "SpatRaster")) {
      if (terra::nlyr(x) != 1) {
        stop(
          "'x' should be a single layer SpatRaster.",
          call. = FALSE
        )
      }
      ext <- terra::ext(x)
      nc <- terra::ncol(x)
      nr <- terra::nrow(x)
      xr <- terra::xres(x) / 2
      yr <- terra::yres(x) / 2
      crs <- st_crs(x)
      lon <- seq(ext[1] + xr, ext[2] - xr, length.out = nc)
      lat <- seq(ext[4] - yr, ext[3] + yr, length.out = nr)
      m <- matrix(
        data = terra::values(x),
        nrow = nr,
        dimnames = list(lat, lon),
        byrow = TRUE
      )
    }
  }


  if (inherits(x = x, what = "sf")) {
    if (missing(var)) {
      stop("'var' is missing.", call. = FALSE)
    }
    if (!var %in% names(x)) {
      stop("'var' is not a valid variable of 'x'.", call. = FALSE)
    }
    crs <- st_crs(x)
    x <- data.frame(st_coordinates(x), var = x[[var]])
    coords <- c("X", "Y")
    var <- "var"
  }


  if (inherits(x = x, what = "data.frame")) {
    if (missing(var)) {
      stop("'var' is missing.", call. = FALSE)
    }
    if (missing(coords)) {
      stop("'coords' is missing.", call. = FALSE)
    }
    if (!var %in% names(x)) {
      stop("'var' is not a valid variable of 'x'.", call. = FALSE)
    }
    if (length(unique(x[[coords[1]]])) * length(unique(x[[coords[2]]])) != length(x[[var]])) {
      stop(
        "It seems that 'x' is not a regular grid.",
        call. = FALSE
      )
    }

    # Reorder dataframe by X-Y if needed
    x <- x[order(x[[coords[2]]], x[[coords[1]]]), ]

    m <- t(
      matrix(
        data = x[[var]],
        nrow = length(unique(x[[coords[1]]])),
        dimnames = list(
          unique(x[[coords[1]]]),
          unique(x[[coords[2]]])
        )
      )
    )
  }

  # breaks management
  vmin <- min(m, na.rm = TRUE)
  vmax <- max(m, na.rm = TRUE)
  if (missing(breaks)) {
    breaks <- seq(
      from = vmin,
      to = vmax,
      length.out = (nbreaks + 1)
    )
  } else {
    breaks <- sort(unique(c(vmin, breaks[breaks > vmin & breaks < vmax], vmax)))
  }

  # isobanding
  lev_low <- breaks[1:(length(breaks) - 1)]
  lev_high <- breaks[2:length(breaks)]
  raw <- isobands(
    x = as.numeric(colnames(m)),
    y = as.numeric(rownames(m)),
    z = m,
    levels_low = c(vmin - 1, lev_low[-1]),
    levels_high = c(lev_high[-length(lev_high)], vmax + 1)
  )

  bands <- iso_to_sfg(raw)
  iso <- st_sf(
    id = 1:length(bands),
    isomin = lev_low,
    isomax = lev_high,
    geometry = st_sfc(bands),
    crs = crs
  )

  # mask mgmt
  if (!missing(mask)) {
    if (inherits(x = mask, "SpatVector")) {
      if (!requireNamespace("terra", quietly = TRUE)) {
        stop(
          paste0(
            "This function needs the 'terra' package to work with ",
            "'terra' objects. Please install it."
          ),
          call. = FALSE
        )
      }
      mask <- st_as_sf(mask)
    }
    st_agr(iso) <- "constant"
    if (st_crs(iso) == st_crs(mask)) {
      iso <- st_cast(st_intersection(x = iso, y = st_union(st_geometry(mask))))
    } else {
      cat(
        "CRS of 'x' and 'mask' should be identical, polygons are not clipped."
      )
    }
  }

  if (sptvct) {
    iso <- terra::vect(iso)
  }


  return(iso)
}
